### ISSUE TITLE
intermittent failure of twisted.web.test.test_webclient.WebClientSSLTestCase.test_downloadTimeout

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3782]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3782|
|Type|defect|
|Created|2009-04-17 13:32:55Z|
|Branch|https://github.com/twisted/twisted/tree/cleanup-after-webclient-timeout-3782|

This seems to happen from time to time on most platforms.

```
[ERROR]: twisted.web.test.test_webclient.WebClientSSLTestCase.test_downloadTimeout

Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<<class 'twisted.internet.tcp.TLSConnection'> to ('127.0.0.1', 46740) at b7ab30c>
<ProtocolWrapper #1 on 46740>
<ProtocolWrapper #0 on 46740>
<<class 'twisted.internet.tcp.TLSConnection'> to ('127.0.0.1', 46740) at b7ab20c>
```

It can be found on buildbot in various places, eg http://buildbot.twistedmatrix.com/builders/debian-easy-py2.5-epoll/builds/386

<details><summary>Searchable metadata</summary>

```
trac-id__3782 3782
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_cleanup_after_webclient_timeout_3782 branches/cleanup-after-webclient-timeout-3782
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__tests tests
time__1239975175000000 1239975175000000
changetime__1241274584000000 1241274584000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [26727]) Branching to 'cleanup-after-webclient-timeout-3782'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

I _think_ this is fixed in the branch.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/cleanup-after-webclient-timeout-3782)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Local tests eventually resulted in a failure:

```
[ERROR]: twisted.web.test.test_webclient.WebClientSSLTestCase.test_timeoutTriggering

Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<<class 'twisted.internet.tcp.TLSConnection'> to ('127.0.0.1', 59660) at 1345af0>
```

This isn't the leak I fixed, but a different one.  Joy.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_4"></a>|@mwhudson commented|
|-|-|

I'm not sure if this branch should still be in review or not?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I filed [#3796](https://github.com/twisted/twisted/issues/3796) for the above mentioned intermittent failure.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Fine, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [26781]) Merge cleanup-after-webclient-timeout-3782 

Author: exarkun
Reviewer: therve
Fixes: [#3782](https://github.com/twisted/twisted/issues/3782)

Fix the HTTP client tests to clean up after themselves more thoroughly so that
they don't leave selectables in the reactor for trial to turn into errors.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
